File: sol/solution/0001-0100/0039. Combination Sum/Solution.cpp
class Solution
{
public:
    vector<vector<int>> res;
    vector<int> comb;
    void dfs(int rootindex, vector<int> &candidates, int target, int sum)
    {
        if (sum == target)
        {
            res.push_back(comb);
            return;
        }
        for (int i = rootindex; i < candidates.size(); i++)
        {
            if (candidates[i] + sum > target)
            {
                continue;
            }
            comb.push_back(candidates[i]);
            sum += candidates[i];
            dfs(i, candidates, target, sum);
            comb.pop_back();
            sum -= candidates[i];
        }
    }
    vector<vector<int>> combinationSum(vector<int> &candidates, int target)
    {
        dfs(0, candidates, target, 0);
        return res;
    }
};

File: sol/solution/0001-0100/0039. Combination Sum/Solution.cs
public class Solution
{
    public IList<IList<int>> CombinationSum(int[] candidates, int target)
    {
        var res = new List<IList<int>>();
        Combination(candidates, 0, target, res, new List<int>());
        return res;
    }
    private void Combination(int[] arr, int ind, int target, List<IList<int>> res, List<int> list)
    {
        if (target == 0)
        {
            res.Add(new List<int>(list));
            return;
        }
        if (ind == arr.Length || target < 0)
        {
            return;
        }
        if (arr[ind] <= target)
        {
            list.Add(arr[ind]);
            Combination(arr, ind, target - arr[ind], res, list);
            list.RemoveAt(list.Count - 1);
        }
        Combination(arr, ind + 1, target, res, list);
    }
}